src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -401,23 +401,38 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
         _HideNotificationIconRequested();
     }
 
+    // We don't want to try to save layouts if we are about to close
+    _getWindowLayoutThrottler.reset();
+    _windowManager.GetWindowLayoutRequested(_GetWindowLayoutRequestedToken);
+
     _window->Close();
 }
 
 LaunchPosition AppHost::_GetWindowLaunchPosition()
 {
-    // Get the position of the current window. This includes the
-    // non-client already.
-    const auto window = _window->GetWindowRect();
+    LaunchPosition pos{};
+    // If we started saving before closing, but didn't resume the event handler
+    // until after _window might be a nullptr.
+    if (!_window)

```

Comment:
It is called in a couple of event handlers that may or may not be on the main thread, so I don't know if I can guarantee it.

Reply:
Its why there is still the try/catch, yes. I admittedly don't have a good sense of the full lifetime of the application, especially as things are closing. I believe in the error that @zadjii-msft found it would be on the main thread (assuming the ui thread is also the one that handles destruction) since it went through `GetWindowLayoutsAsync`.

Reply:
To be more clear, I believe, but have not verified that this is the order of operations in the crash
- Event handler gets called which gives the peasant an `IAsyncOperation`
- The coroutine progresses until it yields on `await resume_foreground ...`
- The application starts getting closed because we called `PostQuitMessage`
- The event loop in `wWinMain` exits so the destructor for `AppHost` is called
- we set `_window = nullptr` and then call `App->Close`. 
- `App->Close()` tries to clear the event handlers / flush any outstanding requests which includes our event handler in (2)
- The event handler resumes and then tries to access `_window` which is now a a nullptr.
- crash

I have attempted to solve the problem in the following ways
- Remove the event handler for `WindowLayoutRequested` before we call `PostQuitMessage` so hopefully there are no outstanding events when we call our destructor
- If we still get into the event handler somehow, we check explicitly if the window is a nullptr so that we don't attempt to use it if it is invalid
- If we fail later (say on usage of `_logic`) then we just wrap everything in a try catch and just ignore the error. I have not seen a case where this fails, but it might just be that any crash here would've crashed on the `_window` usage before.

I think that this is sufficiently resilient to avoiding crashes but it is possible that there is something I'm missing still.